AsyncDisplayKit/Details/ASPhotosFrameworkImageRequest.m

```diff
@@ -0,0 +1,169 @@
+//
+//  ASPhotosFrameworkImageRequest.m
+//  AsyncDisplayKit
+//
+//  Created by Adlai Holler on 9/25/15.
+//  Copyright © 2015 Facebook. All rights reserved.
+//
+
+#import "ASPhotosFrameworkImageRequest.h"
+#import "ASBaseDefines.h"
+
+NSString *const ASPhotosURLScheme = @"ph";
+
+static NSString *const _ASPhotosURLQueryKeyWidth = @"width";
+static NSString *const _ASPhotosURLQueryKeyHeight = @"height";
+
+// value is PHImageContentMode value
+static NSString *const _ASPhotosURLQueryKeyContentMode = @"contentmode";
+
+// value is PHImageRequestOptionsResizeMode value
+static NSString *const _ASPhotosURLQueryKeyResizeMode = @"resizemode";
+
+// value is PHImageRequestOptionsDeliveryMode value
+static NSString *const _ASPhotosURLQueryKeyDeliveryMode = @"deliverymode";
+
+// value is PHImageRequestOptionsVersion value
+static NSString *const _ASPhotosURLQueryKeyVersion = @"version";
+
+// value is 0 or 1
+static NSString *const _ASPhotosURLQueryKeyAllowNetworkAccess = @"network";
+
+static NSString *const _ASPhotosURLQueryKeyCropOriginX = @"crop_x";
+static NSString *const _ASPhotosURLQueryKeyCropOriginY = @"crop_y";
+static NSString *const _ASPhotosURLQueryKeyCropWidth = @"crop_w";
+static NSString *const _ASPhotosURLQueryKeyCropHeight = @"crop_h";
+
+@implementation ASPhotosFrameworkImageRequest
```

Comment:
I you're absolutely right, it's a counterpart to NSURL not NSURLRequest so it'd be ideal to subclass NSURL. Seems like that's not practical since NSURL is immutable. ASPhotosFrameworkURL seems like a bad name though, since really it's URL-convertible not a URL. 

I like your second idea but now I'm thinking it could be billed as some kind of Builder. "This object helps you build and parse ph:// NSURLs" since everyone knows and loves NSURL



Sent from my iPhone

> On Sep 27, 2015, at 12:05 PM, appleguy <notifications@github.com> wrote:
> 
> In AsyncDisplayKit/Details/ASPhotosFrameworkImageRequest.m:
> 
> > +
> > +// value is PHImageRequestOptionsDeliveryMode value
> > +static NSString *const _ASPhotosURLQueryKeyDeliveryMode = @"deliverymode";
> > +
> > +// value is PHImageRequestOptionsVersion value
> > +static NSString *const _ASPhotosURLQueryKeyVersion = @"version";
> > +
> > +// value is 0 or 1
> > +static NSString *const _ASPhotosURLQueryKeyAllowNetworkAccess = @"network";
> > +
> > +static NSString *const _ASPhotosURLQueryKeyCropOriginX = @"crop_x";
> > +static NSString *const _ASPhotosURLQueryKeyCropOriginY = @"crop_y";
> > +static NSString *const _ASPhotosURLQueryKeyCropWidth = @"crop_w";
> > +static NSString *const _ASPhotosURLQueryKeyCropHeight = @"crop_h";
> > +
> > +@implementation ASPhotosFrameworkImageRequest
> This is definitely a cool, powerful class. I wonder if the reason this is tripping me out so much is the billing of it as a "request". I guess it actually is more like an NSURL, itself. I'm not sure if it is permitted to subclass NSURL, but would you agree that would be better conceptually, if it were allowed?
> 
> I hate rename thrash, but it is important we get new APIs right. Maybe this is an ASPhotosFrameworkURL. One option would be to change the return type on the multiplex data source to id, and then aggressively enforce (with exceptions, probably not assertions) that the type be either NSURL or ASPhotosFrameworkURL.
> 
> —
> Reply to this email directly or view it on GitHub.
>